[ZIP-14] Add zip for revised pBFT consensus (#49)
Zilliqa/ZIP: zips/zip-14.md

|  ZIP | Title | Status| Type | Author | Created (yyyy-mm-dd) | Updated (yyyy-mm-dd)
|--|--|--|--| -- | -- | -- |

> | 14  | Revised pBFT consensus | Draft | Standards Track  | Sandip Bhoir <sandip@zilliqa.com>| 2020-12-10 | 2020-12-10

Reviewer: Let's add George Pîrlea george@zilliqa.com to the ZIP

## Abstract

This ZIP details how **Revised pBFT** is adopted to add sustainability to network in high load circumstances where nodes may have varying transactions set.

## Motivation

In the past, the transactions were processed before each node participates in consensus. If nodes misses receiving any of the transactions, it turns out to be too late for him to request the missing transactions from leader and rerun consensus again.
With new pBFT, we handle above issue and thereby adding sustainability enforcing network to progress reliably.

## Specification

### Old pBFT
Please refer below for more details on old behaviour of pBFT for shard microblock.

![existing_flow](../assets/zip-14/existing_flow.png)

### Revised PBFT
Please refer below for more details on new behaviour of pBFT for shard microblock.
Same applies for finalblock consensus.

> ![new_flow](../assets/zip-14/new_flow.png)

Reviewer: ```suggestion
![new_flow](../assets/zip-14/revised_new_flow.png )
```

These are important points to consider:

**Round 1:**

> - Round 1 of consensus also called preprep phase will run consensus on microblock/finalblock with txnhashes only.

Reviewer: ```suggestion
- Round 1 of consensus also known as the `pre-prepare` phase will run consensus on microblock/finalblock with a set of transaction hashes. This phase is for the committee to agree on the order and set of transactions to process. 
```

> - Leader starts actual transaction processing after sending out preprep announcement ( mb/fb with txnhashes only).

Reviewer: ```suggestion
- Leader starts transaction exeuction after sending out `pre-prepare` announcement ( mb/fb with transaction hashes only).
```

> - Backups starts actual transaction processing after sending out commit for preprep phase.

Reviewer: ```suggestion
- Backups start transaction execution after sending out commit to the leader for `pre-prepare` phase.
```

> - However, if there is commit failure ( ex. due to missing txns), it will fetch the missing ones from leader and rerun preprep phase (round-1) of consensus.

Reviewer: ```suggestion
- However, if there is commit failure ( ex. due to missing txns), the backuip will fetch the missing transactions from leader and rerun the `pre-prepare` phase (round-1) of consensus.
```
